# STEPS FOR SIMULATION AND ANALYSIS

We aim to simulate an examination process where two groups of students are compared: one
group uses cheat sheets, and the other does not. We will perform a z-test and adjust parameters
interactively to observe their impact.

## STEP 1

The parameters to simulate the exam process are:
1. Mean score of students using cheat sheets (μ1): range [85, 95], step size: 1
2. Mean score of students not using cheat sheets (μ2): range [80, 90], step size: 1
3. Standard deviation of students using cheat sheets (σ1): range [1, 5], step size: 0.1
4. Standard deviation of students not using cheat sheets (σ2): range [1, 5], step size: 0.1
5. Sample size for both groups (n1, n2): range [20, 200], step size: 10

In [1]:
import numpy as np
from scipy import stats

# Define the parameters
mu1_range = np.arange(85, 96, 1)
mu2_range = np.arange(80, 91, 1)
sigma1_range = np.arange(1, 5.1, 0.1)
sigma2_range = np.arange(1, 5.1, 0.1)
n_range = np.arange(20, 201, 10)

# Example: Simulate for specific values
mu1 = 90
mu2 = 85
sigma1 = 2
sigma2 = 3
n1 = 50
n2 = 50

# Generate random samples
np.random.seed(0)  # For reproducibility
group1_scores = np.random.normal(mu1, sigma1, n1)
group2_scores = np.random.normal(mu2, sigma2, n2)

# Perform z-test
z_stat, p_value = stats.ttest_ind(group1_scores, group2_scores)

print(f"Z-statistic: {z_stat}, P-value: {p_value}")

Z-statistic: 10.874361055604869, P-value: 1.5481893240300766e-18


## Step 2: Calculate Pooled Standard Deviation and Standard Error

Using the provided formulas, calculate the pooled standard deviation (Sp) and standard error (SE)
for each set of parameter values.

In [2]:
# Calculate pooled standard deviation
Sp = np.sqrt(((n1 - 1) * sigma1**2 + (n2 - 1) * sigma2**2) / (n1 + n2 - 2))

# Calculate standard error
SE = Sp * np.sqrt((1 / n1) + (1 / n2))

print(f"Pooled Standard Deviation (Sp): {Sp}")
print(f"Standard Error (SE): {SE}")

Pooled Standard Deviation (Sp): 2.5495097567963922
Standard Error (SE): 0.5099019513592785


## Step 3: Perform Z-test
For each set of simulated parameters, calculate the z-score as, Compare the z-score to the critical value for a one-tailed test (zcrit = 1.645 for α = 0.05).

## Step 4: Compare Z-score to Critical Value

For a one-tailed test with a significance level of α = 0.05, the critical value (zcrit) is 1.645. We compare the calculated z-score to this critical value to determine if we reject the null hypothesis.

- **Z-score (calculated):** 10.874361055604869
- **Critical value (zcrit):** 1.645

Since the calculated z-score is greater than the critical value, we reject the null hypothesis. This indicates that there is a statistically significant difference between the two groups.


## Step 4: Plot Results Using Bar Plots

Plot the z-scores and p-values as the parameters vary. Use interactive sliders to dynamically adjust
the mean, standard deviation, and sample size.

In [10]:
import ipywidgets as widgets
from IPython.display import display

import matplotlib.pyplot as plt

# Function to update the plot based on slider values
def update_plot(mu1, mu2, sigma1, sigma2, n1, n2):
    # Generate random samples
    group1_scores = np.random.normal(mu1, sigma1, n1)
    group2_scores = np.random.normal(mu2, sigma2, n2)
    
    # Perform z-test
    z_stat, p_value = stats.ttest_ind(group1_scores, group2_scores)
    
    # Calculate pooled standard deviation
    Sp = np.sqrt(((n1 - 1) * sigma1**2 + (n2 - 1) * sigma2**2) / (n1 + n2 - 2))
    
    # Calculate standard error
    SE = Sp * np.sqrt((1 / n1) + (1 / n2))
    
    # Clear previous plots
    plt.clf()
    
    # Plot z-score and p-value
    fig, ax1 = plt.subplots()
    
    ax1.set_xlabel('Parameters')
    ax1.set_ylabel('Z-score', color='tab:blue')
    ax1.bar(['Z-score'], [z_stat], color='tab:blue')
    ax1.tick_params(axis='y', labelcolor='tab:blue')
    
    ax2 = ax1.twinx()
    ax2.set_ylabel('P-value', color='tab:red')
    ax2.bar(['P-value'], [p_value], color='tab:red')
    ax2.tick_params(axis='y', labelcolor='tab:red')
    
    fig.tight_layout()
    plt.show()

# Create interactive sliders
mu1_slider = widgets.IntSlider(value=90, min=85, max=95, step=1, description='μ1')
mu2_slider = widgets.IntSlider(value=85, min=80, max=90, step=1, description='μ2')
sigma1_slider = widgets.FloatSlider(value=2, min=1, max=5, step=0.1, description='σ1')
sigma2_slider = widgets.FloatSlider(value=3, min=1, max=5, step=0.1, description='σ2')
n1_slider = widgets.IntSlider(value=50, min=20, max=200, step=10, description='n1')
n2_slider = widgets.IntSlider(value=50, min=20, max=200, step=10, description='n2')

# Display interactive plot
widgets.interactive(update_plot, mu1=mu1_slider, mu2=mu2_slider, sigma1=sigma1_slider, sigma2=sigma2_slider, n1=n1_slider, n2=n2_slider)

interactive(children=(IntSlider(value=90, description='μ1', max=95, min=85), IntSlider(value=85, description='…